File: 2020/day10.js
function day10(){
show('Day10')
let din = loadNumbers('day10.txt').sort((a,b)=>a-b)

let diff = idx => din[idx]-(din[idx-1]||0)
let d = range(din.length).map(diff)
let n3 = d.filter(i=>i==3).length
let n1 = d.filter(i=>i==1).length

show('Part1',n1*(n3+1))

let x = {0:1,1:1,2:2,3:4,4:7}
let it = d.join('').split('3').map(el=>[1,1,2,4,7][el.length])
show('Part2',it.reduce((a,b)=>a*b))
}
day10()

File: 2021/day10.hs
module Main where

import qualified Data.List as L

main = do input <- readFile "day10.txt"
          let l = lines input
              p2 = filter (0 <) (map (part2 []) l)
            in
            do
              print $ sum $ map (part1 []) l
              print $ (L.sort p2) !! (div (length p2) 2)

part1 stack input | input == [] = 0
                  | elem h opening = part1 (h:stack) (tail input)
                  | (head stack == match h) = part1 (tail stack) (tail input)
                  | otherwise = rank h
                  where h = head input

part2 stack input | input == [] = foldl (\acc elem -> acc * 5 + (rank elem)) 0 stack
                  | elem h opening = part2 (h:stack) (tail input)
                  | (head stack == match h) = part2 (tail stack) (tail input)
                  | otherwise = 0
                  where h = head input

opening = "([{<"
match ']' = '['
match ')' = '('
match '}' = '{'
match '>' = '<'
match _ = '_'

rank ')' = 3
rank ']' = 57
rank '}' = 1197
rank '>' = 25137
rank '(' = 1
rank '[' = 2
rank '{' = 3
rank '<' = 4
rank _ = -1000000000

